Deprecate PrimitivesV1 reference implementations and their utils

LeanderCS opened the pull request:
Fixes #12497
### Summary
Deprecate

- BaseEstimatorV1
- BaseSamplerV1
- Estimator
- Sampler
- BackendEstimator
- BackendSampler
- init_circuit function
- init_observable function
- final_measurement_mapping function

Update
- BasePrimitive docstrings

### Details and comments

LeanderCS committed:
Deprecate V1 Primitives and their utils

LeanderCS committed:
Fix tests

LeanderCS committed:
Fix yaml error

LeanderCS committed:
Fix build

LeanderCS committed:
Merge branch 'main' into deprecate-primitives-v1

LeanderCS committed:
Merge branch 'main' into deprecate-primitives-v1

LeanderCS committed:
Fix error after mc

t-imamichi commented on qiskit/primitives/backend_estimator.py:
```diff
@@ -104,6 +105,7 @@ class BackendEstimator(BaseEstimator[PrimitiveJob[EstimatorResult]]):
     precludes doing any provider- or backend-specific optimizations.
     """
 
+    @deprecate_func(since="1.0", additional_msg="Use BackendEstimatorV2 instead.")
```
Thank you for working on this task.

It will be deprecated with the next major release. So, could you update all `since` values and deprecation messages?
```suggestion
    @deprecate_func(since="1.2", additional_msg="Use BackendEstimatorV2 instead.")
```

t-imamichi commented on test/python/primitives/test_estimator.py:
```diff
@@ -346,21 +378,24 @@ class TestObservableValidation(QiskitTestCase):
     )
     @unpack
-    def test_validate_observables(self, observables, expected):
+    def test_validate_observables(self, obsevables, expected):
```
```suggestion
    def test_validate_observables(self, observables, expected):
```

t-imamichi commented on test/python/primitives/test_estimator.py:
```diff
@@ -346,21 +378,24 @@ class TestObservableValidation(QiskitTestCase):
         """Test observables standardization."""
-        self.assertEqual(validation._validate_observables(observables), expected)
+        with self.assertWarns(DeprecationWarning):
+            self.assertEqual(validation._validate_observables(obsevables), expected)
```
```suggestion
            self.assertEqual(validation._validate_observables(observables), expected)
```

LeanderCS committed:
Fix error after mc

LeanderCS committed:
Apply comments

t-imamichi commented on qiskit/primitives/base/base_result.py:
```diff
@@ -42,6 +43,13 @@ def __post_init__(self) -> None:
             ValueError: Inconsistent number of experiments across data fields.
         """
+        warn(
+            "The class ``BasePrimitiveResult`` is deprecated as of qiskit 1.0. "
```
Could you update warning messages as well?
```suggestion
            "The class ``BasePrimitiveResult`` is deprecated as of qiskit 1.2. "
```

LeanderCS committed:
Use correct deprecate version for warning message

t-imamichi commented on qiskit/primitives/base/base_primitive.py:
```diff
@@ -17,11 +17,13 @@
 class BasePrimitive(ABC):
     """Primitive abstract base class."""
 
+    @deprecate_func(since="1.2", additional_msg="Use BasePrimitiveV2 instead.")
```
`BasePrimitiveV2` does not exist.
```suggestion
    @deprecate_func(since="1.2", additional_msg="Use BaseSamplerV2 or BaseEstimatorV2 instead.")
```

t-imamichi commented on qiskit/primitives/base/base_result.py:
```diff
@@ -42,6 +43,13 @@ def __post_init__(self) -> None:
+        warn(
+            "The class ``BasePrimitiveResult`` is deprecated as of qiskit 1.2. "
+            "It will be removed no earlier than 3 months after the release date. "
+            "Use PrimitiveResult class in `qiskit.primitives.containers` instead.",
```
```suggestion
            "Use ``PrimitiveResult`` class in ``qiskit.primitives.containers`` instead.",
```

t-imamichi commented on qiskit/primitives/utils.py:
```diff
@@ -25,8 +25,10 @@
+from qiskit.utils.deprecation import deprecate_func
 
 
+@deprecate_func(since="1.2", additional_msg="There is no replacement for this.")
```
@ElePT I'm wondering what is a good message if there is no replacement. Do you have any idea?
Three functions in this file don't have any replacement.
https://github.com/Qiskit/qiskit/pull/11490/files#r1445895211

LeanderCS replied:
@t-imamichi Would [this](https://github.com/Qiskit/qiskit/pull/12575/files#diff-c789a3af03a0881f870c90d3dcdbd86c3a5a9d293deadbccbaa4c4437953d696R33) variant would be more helpful to the user?

t-imamichi replied:
I looked for replacements and wrote them in additional_msg.

LeanderCS committed:
Update deprecation messages

LeanderCS committed:
Add missed ``

t-imamichi commented on qiskit/primitives/base/base_result.py:
```diff
@@ -46,7 +46,7 @@ def __post_init__(self) -> None:
             "The class ``BasePrimitiveResult`` is deprecated as of qiskit 1.2. "
             "It will be removed no earlier than 3 months after the release date. "
-            "Use PrimitiveResult class in `qiskit.primitives.containers` instead.",
+            "Use ``PrimitiveResult`` class in ``qiskit.primitives.containers` instead.",
```
```suggestion
            "Use ``PrimitiveResult`` class in ``qiskit.primitives.containers`` instead.",
```

t-imamichi commented on qiskit/primitives/utils.py:
```diff
@@ -25,8 +25,13 @@
 
+@deprecate_func(
+    since="1.2",
+    additional_msg="There is no replacement for this. Please adjust your code accordingly and use the new utils.",
```
```suggestion
    additional_msg="Use ``QuantumCircuit.initialize`` instead.",
```

t-imamichi commented on qiskit/primitives/utils.py:
```diff
@@ -45,6 +50,10 @@ def init_circuit(state: QuantumCircuit | Statevector) -> QuantumCircuit:
 
+@deprecate_func(
+    since="1.2",
+    additional_msg="There is no replacement for this. Please adjust your code accordingly and use the new utils.",
```
```suggestion
    additional_msg="Use the constructor of ``SparsePauliOp`` instead.",
```

t-imamichi commented on qiskit/primitives/utils.py:
```diff
@@ -68,6 +77,10 @@ def init_observable(observable: BaseOperator | str) -> SparsePauliOp:
 
+@deprecate_func(
+    since="1.2",
+    additional_msg="There is no replacement for this. Please adjust your code accordingly and use the new utils.",
```
```suggestion
    additional_msg=(
        "Use ``mthree.utils.final_measurement_mapping`` instead. "
        "See https://qiskit-extensions.github.io/mthree/apidocs/utils.html for details."
    )
```

t-imamichi committed:
update releasenote

t-imamichi commented:
@LeanderCS I updated the release note to use the sphinx format. Could you pull the change before working on the update.

Could you deprecate `qiskit.primitives.base.{SamplerResult,EstimatorResult}` too?  I have added them in the updated release note. Both are replaced by `PrimitiveResult`.

t-imamichi committed:
Merge branch 'main' into deprecate-primitives-v1

LeanderCS committed:
Deprecate SamplerResult and EstimatorResult

t-imamichi committed:
fix deprecation warning for SamplerResult and EstimatorResult

t-imamichi commented:
To fix lint errors, I updated `EstimatorResult` and `SamplerResult` to show a deprecation warning using `__post_init__` instead of `__init__`.

LeanderCS commented:
@t-imamichi Sorry, for that back and forth, and thanks for your patience

t-imamichi reviewed (dismissed):
LGTM

1ucian0 commented:
I'm putting this PR `on hold` as there are still open conversations on if or how are we going to do this.

mtreinish commented on qiskit/primitives/utils.py:
```diff
@@ -68,6 +77,11 @@ def init_observable(observable: BaseOperator | str) -> SparsePauliOp:
+@deprecate_func(
+    since="1.2",
+    additional_msg="Use ``mthree.utils.final_measurement_mapping`` instead. "
+    + "See https://qiskit-extensions.github.io/mthree/apidocs/utils.html for details.",
```
A better suggestion here might be to point people to `QuantumCircuit.layout` and `SparsePauliOp.apply_layout` to adjust an operator for a layout instead of tracking permutations with measurements. Although also pointing to `mthree` if they want the same exact functionality does make sense too.

t-imamichi replied:
Thank you for the suggestion. I will add `layout` and `apply_layout` in the message as well as the existing mthree message.

t-imamichi replied:
done

ElePT commented on qiskit/primitives/utils.py:
```diff
@@ -25,8 +25,13 @@
 
+@deprecate_func(
+    since="1.2",
+    additional_msg="Use ``QuantumCircuit.initialize`` instead.",
```
This is a totally minor suggestion, but this function only acted if the input was a `Statevector` instance, so we could maybe be a bit more precise in the message:

```suggestion
    additional_msg="To initialize a circuit from a ``Statevector`` instance, use ``QuantumCircuit.initialize`` instead.",
```

ElePT replied:
This suggestion might require running `black` or fixing "line too long" lint complaints.

t-imamichi replied:
Thanks. I will apply your suggestion and apply black.

t-imamichi replied:
done

t-imamichi committed:
apply review comments

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>
Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

t-imamichi committed:
Merge branch 'main' into deprecate-primitives-v1

ElePT reviewed (dismissed):
LGTM, thanks a lot @LeanderCS for the good work!
Edit: I missed the latest context on the required changes for this PR. I believe the requirements are not met: https://github.com/Qiskit/qiskit/issues/12497#issuecomment-2228393758.

t-imamichi commented:
Oh, I didn't notice the agreement information was in another issue https://github.com/Qiskit/qiskit/issues/12497#issuecomment-2228393758. Thank you, Elena.

t-imamichi committed:
Applying the agreement of deprecations.

- Revert BaseSamplerV1 and BaseEstimatorV1
- Deprecate BaseSampler and BaseEstimator

https://github.com/Qiskit/qiskit/issues/12497#issuecomment-2228393758

t-imamichi commented:
I asked some questions about the agreement https://github.com/Qiskit/qiskit/issues/12497#issuecomment-2246985874. So, let's wait for answers about other base classes and the result classes.

t-imamichi committed:
revert SamplerResult, EstimatorResult, and BasePrimitiveResult

t-imamichi commented:
I reverted SamplerResult, EstimatorResult and BasePrimitive.

t-imamichi committed:
fix test_backend_sampler

t-imamichi committed:
revert tox.ini

t-imamichi committed:
revise deprecation warning for BaseSampler and BaseEstimator

t-imamichi committed:
update reno

ElePT reviewed (commented):
Thanks a lot for the changes @t-imamichi, the deprecated classes seem to follow the recent discussion outcomes. The only missing point is fixing the current lint complaint, which I believe might require modifying the primitives `__init__` to deal with the `BaseEstimator`/`BaseSampler` alias import (probably re-raising the warning in the `__init__`? you may be able to find a better solution than me)

t-imamichi committed:
revert BaseSampler and BaseEstimator

t-imamichi commented:
I thought the revised version of BaseSampler and BaseEstimator was smarter (copied #11054). But as CI showed, it caused complicated import issues. It's not easy for me to resolve because BaseSampler and BaseEstimator are imported by `qiskit.primitives.__init__.py` and `qiskit.primitives.bases.__init__.py`. I don't think it's a good idea to update all init files.
So, I reverted the change. I check pylint locally. I hope CI passes finally.

ElePT reviewed (approved):
LGTM, thanks for the effort taking the PR past the finish line!
